From: Sebastian Ott <sebott@linux.vnet.ibm.com>
Subject: s390/pci: improve handling of hotplug event 0x301
Patch-mainline: v4.2-rc1
Git-commit: 7fc611ff3ff1a0b8f5a6569fe75a97d6c70bed6c
References: bnc#946214, LTC#130628

Description:  s390/pci: fix handling of hotplug events
Symptom:      A pci function added to the system via hotplug
              might not be usable to Linux. In rare situations
              the kernel might crash.
Problem:      Linux does not handle all hotplug events correctly.
Solution:     Handle hotplug events correctly.
Reproduction: pci hotplug

Upstream-Description:

              s390/pci: improve handling of hotplug event 0x301

              Hypervisors may deliver event 0x301 not only for standby
              but also for reserved devices.
              Just handle event 0x301 regardless of the device's state.

              Reviewed-by: Gerald Schaefer <gerald.schaefer@de.ibm.com>
              Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Sebastian Ott <sebott@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 arch/s390/pci/pci_event.c |    8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

--- a/arch/s390/pci/pci_event.c
+++ b/arch/s390/pci/pci_event.c
@@ -75,7 +75,13 @@ static void __devinit __zpci_event_avail
 	zpci_err_hex(ccdf, sizeof(*ccdf));
 
 	switch (ccdf->pec) {
-	case 0x0301: /* Standby -> Configured */
+	case 0x0301: /* Reserved|Standby -> Configured */
+		if (!zdev) {
+			ret = clp_add_pci_device(ccdf->fid, ccdf->fh, 0);
+			if (ret)
+				break;
+			zdev = get_zdev_by_fid(ccdf->fid);
+		}
 		if (!zdev || zdev->state != ZPCI_FN_STATE_STANDBY)
 			break;
 		zdev->state = ZPCI_FN_STATE_CONFIGURED;
